### ISSUE TITLE
[discussion:general] What should be a template, image, feature, postCreateCommand?

### ISSUE THREAD

**Issue by jcbhmr** [devcontainers-contrib/.github#4]
**This should be an org-wide discussion.** Once Discussions are enabled on this org/repo, move this issue to a Discussion pls @danielbraun89.

I am creating this Discussion to centralize discussion on this particular topic. This topic of "what makes a feature a feature?" and "what belongs in a feature and what goes in a postCreateCommand?" has been discussed in https://github.com/devcontainers-contrib/features/discussions/171, https://github.com/devcontainers-contrib/features/discussions/173, and now https://github.com/devcontainers-contrib/features/discussions/202. This is a centralization of those questions.

To elaborate on the question, we want to address this particular example scenario: (If you have another wrench to throw on the pile, feel free to do so; this is a contrived example that attempts to cover most edge cases)

> A user is writing a [Python] project that uses [Prettier] to format the README.md file. They use [Poetry] to manage dependencies for this Python project. Their API documentation is generated using [Sphinx]. They use GitHub Actions to build, test, and publish the project. They also want to use [poethepoet] for task running.

[python]: https://www.python.org/
[prettier]: https://prettier.io/
[poetry]: https://python-poetry.org/
[sphinx]: https://www.sphinx-doc.org/en/master/
[poethepoet]: https://github.com/nat-n/poethepoet

**Open questions** (these are the ones that warrant discussion)

1. Which devcontainer template should they use to create their initial devcontainer.json file?
2. Which devcontainer image should they use as a base?
3. What features should they use and what options should they pass to them?
4. What should go in the postCreateCommand?

My own (@jcbhmr's) opinion on this is posted as a comment to this Discussion.

**Comment by jcbhmr** [comment]
_In my opinion_, this seems like it should have the following devcontainer.json (reasoning below)
```json
{
  "image": "ghcr.io/devcontainers-contrib/images/python-with-poetry",
  "features": {
    "ghcr.io/devcontainers/features/node": {}
  },
  "postCreateCommand": "./configure && poetry install"
}
```

The real key here is the...
1. The use of a python-with-poetry image
2. Lack of options passed to the `node` feature
3. The use of an [autotools]-like `./configure` script[^1] to do magic ✨ (see below for contents)
4. The `poetry install` being in the postCreateCommand, NOT in `./configure`

## 1. The use of a python-with-poetry image

The reason I'm using an example (non-existant as of yet) python-with-poetry image is to demonstrate that images should be designed to meet the requirements of what a typical developer's normal "devcomputer" (their typical working computer) has installed on it for their workflow. In this case, a Python-with-poetry dev can be expected to have the assorted Python, pip, poetry, GNU core utils, etc. all installed.

Here are some images that don't exist now (not in the official devcontainers/images repo) but that I think should exist to fit other workflows:
- `python-with-poetry`
- `python-with-cpp`
- `node-with-yarn`
- `node-with-pnpm`
- `cpp-with-conan`
- `cpp-with-vcpkg`
- `ubuntu-with-homebrew`
- `haskell`
- `java-with-maven`

Notice how all of these images are like typical dev environments? There are certain sub-images within scopes; like Node.js has plain old Node.js, node+yarn and node+pnpm. Same with Python. Python has plasin Python, Python+Poetry, and maybe Python+cpp for developing native Python extensions (I'm no expert, I just wanted another Python offshoot example).

## 2. Lack of options passed to the `node` feature

Now lets address what I think makes a feature a feature. Features should act almost exactly like images. They should have a minimal amount of configuration options; likely just versions, similar to how images work:

```
mcr.microsoft.com/devcontainers/python:3 (latest)
mcr.microsoft.com/devcontainers/python:3.7 (or 3.7-bullseye, 3.7-buster to pin to an OS version)
mcr.microsoft.com/devcontainers/python:3.8 (or 3.8-bullseye, 3.8-buster to pin to an OS version)
mcr.microsoft.com/devcontainers/python:3.9 (or 3.9-bullseye, 3.9-buster to pin to an OS version)
mcr.microsoft.com/devcontainers/python:3.10 (or 3.10-bullseye, 3.10-buster to pin to an OS version)
mcr.microsoft.com/devcontainers/python:3.11 (or 3.11-bullseye, 3.11-buster to pin to an OS version)
```

(https://github.com/devcontainers/images/tree/main/src/python)

Each of these features should represent the same concept of a "toolchain" that a root image does. For instance, in this example from the Discussion, a Python developer needs to use the Node.js toolchain. So what do they do? They add a Node.js feature to their basic Python image since there is no python-with-nodejs image for a pre-configured devcontainer yet.

**If an image repeatedly sees the same feature used with it over and over, that's the sign a new sub-image should be created!** For instance, if this Python & Node.js dev environment became a new trend it should get a `python-with-node` image made to make it easier!

Now back to the "Lack of options passed to the `node` feature" thing... These features should act like "default toolchains" as though someone just happened to have two of them installed. Basically what I'm saying is that features should act like the contents of the Dockerfile were pulled into an `install.sh` script instead of being tied to a container. The following should be identical with **no extra options in the feature** that aren't present on the default image.

<table>
<tr><td>

```json
{
  "image": "mcr.microsoft.com/devcontainers/python"
}
```

<td>

```json
{
  "image": "mcr.microsoft.com/devcontainers/base",
  "features": {
    "ghcr.io/devcontainers/features/python": {}
  }
}
```

</table>

Adding an option like one described in https://github.com/devcontainers-contrib/features/discussions/202 would break that contract and add things to features that _aren't present by default_ and critically **cannot be expected to be that way on an average developer's laptop**. Instead of being in an easy-to-see shell script, these `sudo poetry self add poethepoet[poetry_plugin]` commands are now stuffed in an obscure devcontainer feature image that a non-devcontainer user isn't going to know what (below) means!

```jsonc
{
  "features": {
    "ghcr.io/devcontainers-contrib/poetry": {
      // You'll have to repeat these instructions in the readme to make sure
      // that non-devcontainers users do this too!
      "plugins": "poethepoet[poetry_plugin]"
    }
  }
}
```

Compare that to a common centralized setup.sh or postCreateCommand that handles these non-default things! That leads into...

## 3. The use of an [autotools]-like `./configure` script[^1]

This is where we do the addition of poethepoet task runner and install prettier.

```sh
#!/bin/bash -e
npm i -g prettier
poetry self add poethepoet[poetry_plugin]
```

The biggest pro of this method over using feature options is that users who don't use devcontainers who have the needed Node.js and Python stuff installed can just run `./configure` and it will do it. If instead you put those options into the devcontainer.json, now those people who don't use devcontainers (**like your CircleCI service!**) are fresh out of luck and you now need to write those shell commands _anyways_. You might as well use that configure script for two things instead of one.

Thus this:
1. De-deuplicates specifying poetry plugins in the devcontainer.json and in a CI configure script
2. Allows non-devcontainers users an easier onboarding experience for non-standard customizations (like global pip or npm modules and custom poetry plugins)
3. Explicitly allows those non-devcontainers users to see _exactly what they need to undo_ after they're done with the project (`npm uninstall -g prettier` etc.)

## 4. The `poetry install` being in the postCreateCommand, NOT in `./configure`

Honestly, this was just a stylistic choice because poetry has a CI mode. Thus, in CI it's easy to run the (below) to de-couple the poetry installation from the ./configure setup.

```yml
- run: ./configure
- run: poetry install --without dev
```

If `./configure` had done `poetry install`, then it would have downloaded those dev deps that we don't need. npm has a similar thing with `npm install` vs `npm ci` which determines if it install dev deps or not. Ideally, those should be separate; maybe even excluded from that postCreateCommand altogether, depending on your purism!

---

Anyways, that's my opinion on the solution and architecture to this issue. Thoughts? Comments? Counter-arguments?

/cc @danielbraun89 @DaneWeber (if you're still interested; tell me if you want me to stop pinging you about these things) and @imaxerik

[^1]: I've lately been inspired to use `./configure` as my go-to setup script, but `tools/setup.sh` or `install.sh` also work too.

[autotools]: https://www.gnu.org/software/automake/manual/html_node/GNU-Build-System.html#:~:text=Compiling%20his%20package%20was%20now%20as%20simple%20as%20running%20./configure%20%26%26%20make.

**Comment by imaxerik** [comment]
I'm just starting to familiarize myself with the way Dev Containers and features work so I don't have (m)any opinions yet, but I do like what I've read so far in these discussions.

There's one thing that stood out to me, though:

> **If an image repeatedly sees the same feature used with it over and over, that's the sign a new sub-image should be created!** For instance, if this Python & Node.js dev environment became a new trend it should get a `python-with-node` image made to make it easier!

I'd use caution with that, and maybe recommend a depth limit, lest sub-image names start giving isekai web novel titles a run for their character counts. For example, both `python-with-poetry` and `python-with-pipenv` (pipenv is similar to poetry) would likely be useful, but `python-with-poetry-and-black` and `python-with-pipenv-and-black` might be straying into silly territory (black is a Python code formatter). Some commonly used features may be better off as features that are easily composed together rather than being baked into images.

**Comment by jcbhmr** [comment]
/re @imaxerik 

> For example, both python-with-poetry and python-with-pipenv (pipenv is similar to poetry) would likely be useful, but python-with-poetry-and-black and python-with-pipenv-and-black might be straying into silly territory (black is a Python code formatter).

Good points. I'd say it's kinda discretionary. Features are like "add-ons". If an addon becomes so popular, it might good to bake it in to your product/thing. I do agree that the names could get silly. At some point, yeah you'd have to cut it off.

Since there's only likely to be _maybe_ 4-6 truly "mainstream" configs, I don't think this is _too_ much of an issue.

Also quick question about black: it seems kinda like it should be what I (from the npm world) think of as a dev dependency that is listed in the package.json, kinda like prettier is for nodejs projects. Does python have a mechanism to that too, or are dev-only packages installed globally?

```json
"scripts": {
  "lint": "prettier -w ."
},
"devDependencies": {
  "prettier": "latest" 
}
```

Some quick googling suggests that poetry has [dependency groups](https://python-poetry.org/docs/master/managing-dependencies/) but I'm wondering if that's actually used in the real world. i.e. have you seen it?

**Comment by danielbraun89** [comment]
I would like to offer another point of view,  not because im sure its correct (im not), but in order to enrich the discussion 😄 

Lets look at the devcontainer as a result of a configuration "evolution" if you will.
So we begin with a custom docker file and some configuration line. looks something like this:
```json
{
  "build": { "dockerfile": ".Dockerfile" },
  "postCreateCommand": "./configure && poetry install"
}
```
and the dockerfile is something like this:

```dockerfile
FROM ubuntu:22.04
# install python
# install poetry
```

This had the problem of rewriting or copy pasting the same dockerfile to every other project or writing it from scratch. no comunity effort.

Enter custom images  `python-with-poetry` `python-with-cpp`  `node-with-yarn` `node-with-pnpm` ,  variants on these images already exists and already flooding docker hub. I know I personally also built something similar to my project, that was to avoid the build time but It was a mistake. These suffer the issues of:

1. Tend to be bloated, in fear of missing some important component so usually they end up difting towards being catch-all universal image
2. They tend to be outdated fast, this is because they stuck in whatever base image they started from forever (ubuntu:18 for example) 
3. The images are indeed communal, but the layers on the dockerfile which originate it arent.  each docker image author craft his own docker layers and install things differently, so no community benefit there.


I feel at this point, devcontainer "features" is the next evolutionary step, in the sense the solve all of the above problems. lets see how:
1. Feature selection happens dynamicly means you choose what you want when you need it, no need for bloat.
2. Features are decoupled from the base image, thus enabling you to update uubntu18 to ubuntu22 for example without changing anything else.
3. Lastly - features can be seen as common and sharable docker layers, with the additional benefit of having a standard interface. no longer copy pasting docker layers from some online example/stackoverflow.

So I feel we are going a step backwards in a sense, giving up on these benefits when using custom image like `ghcr.io/devcontainers-contrib/images/python-with-poetry`.

optimal might be something like 

```json
{
  "image": "debian",
  "features": {
       "python": {},
       "poetry": {},
  "postCreateCommand": "./configure && poetry install"
}
```

I concur poetry as a feature is itself a decision that I have to defend. As you said it can easily be seen as parameter to the python feature, or a `postCreateCommand` line. I guess this should be seen as case to case basis

The easiest justification for a feature is if its installation is non trivial, yet common enough that its worth packaging it in a sharable way. Another reason might be that it has installation gotchas that you want to help people avoid. But again thats easy
Poetry, as an example, falls in that category exactly because it offers some more intricate installation/configuration (the plugins) so you might add it to its interface and save everybody's time. 
( btw, In the future I think each company would probably maintain its own product features. so Poetry for example has a for-profit company behind it so they should maintain the feature for their product and update it wit new flags and installation params etc. This should scale the best)

There is another category of feature reasoning - availability to a _non-experienced user in that ecosystem_ 
So for example say you are js developer and you need some python command line tool ( for example cookiecutter.  ansible. etc). You dont feel comfortable enough with python to install and manage it and install a package etc, so its easier for you to use a feature for that. Or say a python developer who is not comfortable with the javascript ecosystem, but want to quickly use  the fkill command line program. Yes, Im not sure about this one as well... 

Regarding templates, I fully agree about 1 to 1 relation between template and custom images now that I think of it.  And because templates have well defined structure they are so much more customizable / maintainable / explanatory  they are preferable. maybe they an supersede dockerfiles as a means to build production images as well? one can dream 😛

**Comment by jcbhmr** [comment]
/re @danielbraun89 

I do agree with some of your sentiments! Thanks for providing another opinion. It's nice to have someone else point out problems that I didn't see.

In your arguments to use features instead of images, you mention that "Feature selection happens dynamicly means you choose what you want when you need it, no need for bloat."

I think this can lead to a _different_ kind of bloat: build-time bloat. For instance, the `mcr.microsoft.com/devcontainers/universal` image is lightning fast. If you start a Codespace on a random repo, it's ready in 15 seconds flat.

Compare that to a .devcontainers/devcontainer.json that looks like this:

```json
{
  "image": "mcr.microsoft.com/devcontainers/base",
  "features": {
    "ghcr.io/devcontainers/features/python": {},
    "ghcr.io/devcontainers/features/node": {},
    "ghcr.io/devcontainers/features/java": {},
    // ...
  }
}
```

Each time you start a new devcontainer, it takes _literally 10 minutes_ to get it built. It has become a Dockerfile. Sure, you can cache it, but that is, in essence **just turning a devcontainer.json into a container Docker image, just like a Dockerfile would**.

What I'm getting at is that Docker images in the `image` key can be cached and pre-made to make them faster. Features are add-ons that are _dynamically executed_ as your devcontainer comes online. In my brain 🧠 they're sort of like `postCreateCommand`, but on steroids and in `install.sh` script.

**TL;DR:** Features (in my experience) lead to poor dev onboarding experience from long build times. Images provide better, faster, _cached_ common environments.

> There is another category of feature reasoning - availability to a non-experienced user in that ecosystem
>
> So for example say you are js developer and you need some python command line tool ( for example cookiecutter. ansible. etc). You dont feel comfortable enough with python to install and manage it and install a package etc, so its easier for you to use a feature for that. Or say a python developer who is not comfortable with the javascript ecosystem, but want to use fkill or heck devcontainer cli. Yes, this is much weaker argument, but still worth mentioning

I actually do agree that this is a sore spot. It is difficult to incorporate things you know nothing about. devcontainer features _could_ become a "common denominator", but I personally find the idea of maintaining _so many_ features an impossible task. It would also create a "should I use the pre-made feature, or add it to my `requirements.txt` file?" dilemma for users. Obviously, if you're in a Python project you should desire to put your deps in a `requirements.txt` file so that others can install the project, but "it's easier just to add it to the devcontainer features" will likely prove persuasive.

**TL;DR:** I agree that for some things like `black` or other _super common_ global tools (maybe `prettier` or `tsc` too, idk) there is _an argument_ to be made to put them in features. I just also think that it pulls those dev dependencies out of the normal flow of the package.json or requirements.txt and puts them somewhere where they shouldn't be: globally.

> * Tend to be bloated, in fear of missing some important component so usually they end up difting towards being catch-all universal image
> * They tend to be outdated fast, this is because they stuck in whatever base image they started from forever (ubuntu:18 for example)
> * The images are indeed communal, but the layers on the dockerfile which originate it arent.  each docker image author craft his own docker layers and install things differently, so no community benefit there.

I agree that these are tough issues.

> ![image](https://user-images.githubusercontent.com/61068799/208319092-07851a22-5384-416d-942c-40f16886498e.png)
>
> -- https://xkcd.com/754/

Ideally, all three of  these issues could  be solved with good governance, similar to how the official devcontainer images work right now: You need a _very_ good reason to add one, and a _very_ good reason to change stuff about it.

More specifically:
1. Good governance from people like you who see the evils of a "universal" container could stem the "add one more feature" (hopefully)
2. I don't know exactly how to solve this one, but probably regular dependabot pings? Again, good maintainers and governance would _help_, but yeah this is a tough problem. You're literally fighting time.
3. Ideally the installation method wouldn't matter. I know there's some discussion in https://github.com/devcontainers-contrib/features/discussions/173 about asdf as a mainline dep manager, and how features could be installed from multiple sources... **BUT** I don't think it matters where the thing was installed from, so long as that when the user logs in to their VS Code instance they can run `deno run my-app.js` or `python3 my-script` and have it _work_. Basically what I'm saying is that installation doesn't matter, only `$PATH` exposure matters.

**Comment by jcbhmr** [comment]
@danielbraun89 I don't know if this is a good idea or not, but we can always create _more_ repositories to hold other "types" of features. For instance, there seems to be a clear-ish divide between features that are "ecosystem" or "expected" (common on devcomputers) and those that are `npm i -g @aws-amplify/cli` in disguise. We can always make another repo to put a more concrete divide or separation between such features:

- devcontainers-contrib/features
- devcontainers-contrib/ez-features
- devcontainers-contrib/ecosystem-images
- devcontainers-contrib/ecosystem-features
- devcontainers-contrib/npm-features
- devcontainers-contrib/github-features
- devcontainers-contrib/popular-features

Let me know what you think. I'm not 100% behind this idea, but it's a possibility that I've warmed to somewhat.
